server/src/Dao/AbstractDao.php

```diff
@@ -2,113 +2,76 @@
 
 namespace App\Dao;
 
+use App\Service\ConnectionServiceInterface;
 use Psr\Log\LoggerInterface;
-use Psr\Log\NullLogger;
+use Doctrine\DBAL\Query\QueryBuilder;
 
 abstract class AbstractDao {
-
-	/**
-	 * @var NullLogger
-	 */
-	protected $logger;
-
 	/**
-	 * @var array
+	 * @var \Doctrine\DBAL\Connection
 	 */
-	protected $config;
+	protected $conn;
 
 	/**
-	 * @var array [dsn => pdo]
+	 * @var
 	 */
-	protected $connections = [];
+	protected $logger;
 
 	/**
-	 * AbstractDao constructor.
-	 *
-	 * @param array $config
-	 * @param LoggerInterface|null $logger
+	 * @param ConnectionServiceInterface $connectionService
+	 * @param LoggerInterface $logger
 	 */
-	public function __construct( $config, LoggerInterface $logger = null ) {
-		$this->logger = $logger ?: new NullLogger();
-		$this->config = $config;
+	public function __construct(
+		ConnectionServiceInterface $connectionService, LoggerInterface $logger
+	) {
+		$this->conn = $connectionService->getConnection();
+		$this->logger = $logger;
 	}
 
 	/**
-	 * @param string $projectName
-	 * @return \PDO
+	 * @param QueryBuilder $qb
+	 * @param string $key
+	 * @param string $sortDir
+	 * @param string $continue
+	 * @param int $fetchMode
+	 * @return array [ results, continue ]
+	 * @throws \Exception
 	 */
-	public function getDb( $projectName ) {
-		$dsn = $this->buildDsn( $projectName );
+	public function paginate(
+		QueryBuilder $qb, $key, $sortDir, $continue, $fetchMode = \PDO::FETCH_ASSOC
+	) {
+		if ( $continue ) {
+			$indexKey = base64_decode( $continue );
+			$sortDir = ( $sortDir == 'asc' ) ? '>=' : '<=';
+			$qb->andWhere( $key . ' ' . $sortDir . ' :indexKey' )
+				->setParameter( ':indexKey', $indexKey );
+		}
 
-		if ( isset( $this->connections[$dsn] ) ) {
-			$pdo = $this->connections[$dsn];
+		$this->logger->debug( $qb->getSQL(), $qb->getParameters() );
+
+		$limit = $qb->getMaxResults();
+		$qb->setMaxResults( $limit + 1 );
+
+		$stmt = $qb->execute();
+		$results = $stmt->fetchAll( $fetchMode );
+
+		// if we have limit + 1, create continue token and remove extra row
+		if ( count( $results ) > $limit ) {
+			$lastRow = array_pop( $results );
+			switch ( $fetchMode ) {
+				case \PDO::FETCH_ASSOC:
+					$continue = base64_encode( $lastRow[$key] );
```

Comment:
oh! sorry, another thing. why base64 encode/decode the continue? why not just pass the id?

Reply:
Having base64 will keep it consistent looking even when paginating different tables with different keys.
Also, initially I thought on maybe adding some hash of the query itself for validating the `continue` param but I ended up not going that route.